### ISSUE TITLE
mock 1.1.1 breaks on python 2.7.3

### ISSUE THREAD

**Issue by xrotwang** [testing-cabal/mock#262]
The [import of `funcsigs` in `mock.py`](https://github.com/testing-cabal/mock/blob/master/mock/mock.py#L83) does not work with python 2.7.3:

```python
Python 2.7.3 (default, Jun 22 2015, 19:33:41) 
[GCC 4.6.3] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import mock
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/robert/venvs/testmock/local/lib/python2.7/site-packages/mock/__init__.py", line 2, in <module>
    import mock.mock as _mock
  File "/home/robert/venvs/testmock/local/lib/python2.7/site-packages/mock/mock.py", line 83, in <module>
    import funcsigs
ImportError: No module named funcsigs
```

with these packages:

```
(testmock)robert@astroman:~/venvs/testmock$ pip freeze
argparse==1.2.1
distribute==0.6.24
mock==1.1.1
pbr==1.2.0
six==1.9.0
wsgiref==0.1.2
```

**Comment by rbtcollins** [comment]
duplicate of #261

**Comment by asottile** [comment]
This is a real issue, you can't expect everyone to be on latest setuptools (in fact latest setuptools breaks other parts of our build).

**Comment by rbtcollins** [comment]
So thats fair enough: you've pinned setuptools back in your environment to deal with some issue or other: you need to pin mock back too, until you're ready to roll forward.

**Comment by rbtcollins** [comment]
(or manually add funcsigs to your dependencies)

**Comment by asottile** [comment]
You shouldn't make breaking changes in a patch release either.

**Comment by rbtcollins** [comment]
You can argue that depending on features from another library is a breaking change, but by that logic every dep change is a major release: an assertion that I reject.

Anyhow, I don't expect everyone to be on latest setuptools - 17.1 should be sufficient (which is quite new, yes). But it doesn't make sense to me that you'd be upset that when one of your deps (setuptools) is capped, that you have to cap other ones as well.

**Comment by asottile** [comment]
I'm upset because I cannot use tox + virtualenv which are installed at the system level and outside of my control with your package because your setup.py has a bug in it.  There are plenty of other packages which have conditional dependencies which do not suffer from the same issue.

**Comment by rbtcollins** [comment]
There aren't that can build universal wheels. conditional dependencies expressed as procedural code are not evaluated in wheels and thus require per-version (and in this case highly specific) wheels.

OR, we can use the stuff the Python packaging ecosystem has been hard at work making a reality.

**Comment by asottile** [comment]
the sdist working would be an improvement, no?

**Comment by asottile** [comment]
The other complaint I have is the versions you are targeting aren't going to have this module in the usual case and people on python3 are just going to use unittest.mock.  would it not make sense to invariantly require funcsigs or at the least partition out a py2 wheel and a py3 wheel.

**Comment by rbtcollins** [comment]
So, mock 1.1 has features not present in unittest.mock from python 3.2, 3.3, 3.4. I fully expect folk to just use mock everywhere, except-and-unless they are a) writing code to go in the stdlib itself, or b) writing code for one and only one version of Python (and even then, unless its 3.5, they'll benefit by using mock rather than unittest.mock).

**Comment by rbtcollins** [comment]
Also, no. funcsigs is needed in Python 3.2 installs.

**Comment by petarmaric** [comment]
First of all, please stop with the version spams (8 mock releases in less than a month) and please reserve all the backwards incompatible changes to MAJOR mock versions, i.e. 2.x, 3.x, ...

People have certain expectations when dealing with libraries, like not breaking when they upgrade from A.x to A.x+1.

And can you at least please provide us with a changelog? In https://github.com/testing-cabal/mock/commit/18c9bbd7e67055041260ae905891dee74586409c you've **removed the existing** changelog and are now referencing a nonexistent file in https://github.com/testing-cabal/mock/blob/1.3.0/docs/changelog.txt.

**Comment by berkerpeksag** [comment]
Hi @petarmaric, changelog is here: https://mock.readthedocs.org/en/latest/changelog.html Perhaps we should add a link to README and PyPI to make it more visible. See also #299 for a similar discussion about changelog.

**Comment by petarmaric** [comment]
Thank you @berkerpeksag, I've read the linked changelog. Some thoughts:
* The changelogs aren't very informative, see 1.2.0 and 1.3.0 changelogs and try to figure out what happened **just by reading the text**
* Why are you not using semantic versions? For example, 1.0.1 users got backwards incompatible changes in 1.1.x. Furthermore, both 1.2.0 and 1.3.0 look a lot more like patch level releases to me, judging from the changelogs.
* What's with the release spam? So many versions, much wow?
* Why is the changelog not stored in the source code repository? AFAIK it's not considered best practice to store important information related to the project, outside the context of the project
* Why is the bleeding edge forced upon the users? Some of us are very happy (or have to) use the software as packaged by the latest Ubuntu LTS release

Love the project, just don't understand what happened to it during July 2015...

**Comment by berkerpeksag** [comment]
> The changelogs aren't very informative, see 1.2.0 and 1.3.0 changelogs and try to figure out what happened just by reading the text

Agreed. I find https://github.com/testing-cabal/mock/blob/master/NEWS more informative. Perhaps we can find a better way.

> What's with the release spam? So many versions, much wow?

Because the development model of mock has slightly changed and if I remember correctly,  we've introduced a couple regressions. I personally prefer to be in-sync with stdlib versions of modules(the alternative is having a bunch of diverged modules such as argparse, simplejson/json, sqlite etc).

For your other questions, I think @rbtcollins can give you much more comprehensive answers :)

**Comment by rbtcollins** [comment]
https://mock.readthedocs.org/en/latest/index.html#library is where NEWS is rendered. Making that have version boundaries would be good I think, if we can make it reasonably low overhead to maintain.

The changelog (https://mock.readthedocs.org/en/latest/changelog.html) is derived from the commit messages - both those in CPython - and those locally. The only local changes we have now are those required to backport. I'm not sure its worth making the backport process substantially harder to add more prose around the commits from CPython.

We are using semver: we just didn't realise that 1.1.0 was backwards incompatible. We haven't introduced new incompatibilities since then. The incompatibilities in 1.1.0 were all done in CPython itself - and not ever recognised as incompatible there. See the centi thread that erupted when folk realised that typos were being handled, for instance.

We had a number of versions to deal with the consequences of the incompatibilities in 1.1.0, and then the poor detection of incompatible setuptools versions.

The changelog is maintained in the source repository, using the git changelog - which is a best practice.

I presume your bleeding edge comment is about the setuptools dependency? Setuptools in Ubuntu LTS precise is incapable of permitting correct universal wheels to be built; since we had to force a newer version there to do that, we had a choice of not requiring something modern, or making that requirement - since we were making it, we chose a good modern version that delivers setuptools-aware conditional dependencies, permitting correct universal wheels to be built.

IMO if folk want to use mock from PyPI, they are opting into using the ecosystem rather than distribution packages. Its fine to use distribution packages, or its fine to use the Python ecosystem, but trying to mix and match between them puts a huge anchor on the ecosystem evolution - decades of delay - and while we need to be graceful and cautious about compatibility, delaying any infrastructure changes by decades each time is unsustainable.
